src/librustc/ty/structural_impls.rs

```diff
@@ -839,14 +839,17 @@ impl<'tcx> TypeFoldable<'tcx> for ty::ExistentialProjection<'tcx> {
 
 impl<'tcx> TypeFoldable<'tcx> for ty::ProjectionTy<'tcx> {
     fn super_fold_with<'gcx: 'tcx, F: TypeFolder<'gcx, 'tcx>>(&self, folder: &mut F) -> Self {
+        let folded_trait_ref = ty::tls::with(|tcx| self.trait_ref(tcx).fold_with(folder));
         ty::ProjectionTy {
-            trait_ref: self.trait_ref.fold_with(folder),
+            // FIXME(tschottdorf): remove once the field is dropped.
+            deprecated_trait_ref: folded_trait_ref,
+            substs: folded_trait_ref.substs,
             item_def_id: self.item_def_id,
         }
     }
 
     fn super_visit_with<V: TypeVisitor<'tcx>>(&self, visitor: &mut V) -> bool {
-        self.trait_ref.visit_with(visitor)
+        ty::tls::with(|tcx| self.trait_ref(tcx).visit_with(visitor))
```

Comment:
Oh, right, visitor. You'll have to update all the implementers that overload `visit_trait_ref` to handle `TyProjection` themselves. At this rate I doubt `visit_trait_ref` is useful anymore since it shouldn't appear in a `Ty`. What you have to do here fwiw is to visit only substitutions.